Metrics API Scaler support different formats

Friedrich42 opened the pull request:
<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md
-->

_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

<!--
  Make sure to link the related issue for this change
  If it requires multiple PRs and/or a PR on another repo as well, please use "Relates to" instead.
-->
Fixes #2633 

<!--
  Make sure to link the related PRs for changes such as documentation & Helm charts
-->

zroubalik commented:
Do we have any update on this please?

murr-rs-42 commented:
@zroubalik 
It is almost ready, but I need a quick review to check if this way of solving that issue is acceptable.

If it's all good I'll finish this PR using the approach I have described in the issue.

https://github.com/kedacore/keda/issues/2633#issuecomment-1903771761

Friedrich42 commented:
It seems like I have messed up the history, because I pulled the changes from kedacore:main and then tried to add DCO. I'll fix it asap

Friedrich42 commented:
Hey. Hope you are well. Could you please review my code? I think it's pretty much ready.
If all is good, I'll finish the rest of the checklist and it's ready to be merged.
@zroubalik

zroubalik reviewed (commented):
There are test failures:

```
Test failures:
  TestGetValueFromResponse: github.com/kedacore/keda/v2/pkg/scalers
  TestGetValueFromResponse/json:_string: github.com/kedacore/keda/v2/pkg/scalers

    metrics_api_scaler_test.go:155: Expected 64, got 64

  TestGetValueFromResponse/json:_{}.[].{}: github.com/kedacore/keda/v2/pkg/scalers

    metrics_api_scaler_test.go:155: Expected 32, got 32

  TestGetValueFromResponse/yaml:_integer: github.com/kedacore/keda/v2/pkg/scalers

    metrics_api_scaler_test.go:151: Expected success but got error: yaml: mapping values are not allowed in this context
    metrics_api_scaler_test.go:155: Expected 2.43, got 0

  TestGetValueFromResponse/yaml:_string: github.com/kedacore/keda/v2/pkg/scalers

    metrics_api_scaler_test.go:151: Expected success but got error: yaml: mapping values are not allowed in this context
    metrics_api_scaler_test.go:155: Expected 64, got 0

  TestGetValueFromResponse/yaml:_{}.[].{}: github.com/kedacore/keda/v2/pkg/scalers

    metrics_api_scaler_test.go:151: Expected success but got error: yaml: mapping values are not allowed in this context
    metrics_api_scaler_test.go:155: Expected 32, got 0
    ```

This should be also covered by [e2e tests](https://github.com/kedacore/keda/blob/main/tests/README.md):

https://github.com/kedacore/keda/blob/main/tests/scalers/metrics_api/metrics_api_test.go

Thanks!

Friedrich42 commented:
@zroubalik I have fixed the tests, but I would need some guidance with E2E. Where can I read about that?

zroubalik commented:
/run-e2e metrics-api

zroubalik commented:
/run-e2e metrics_api

Friedrich42 commented:
The tests seem to pass. Shall I also document this change in keda-docs?

Friedrich42 commented:
Documentation is done as well. You can check it here:
https://github.com/kedacore/keda-docs/pull/1325

Friedrich42 commented:
I believe that it can be merged if you don't have any additional comments.
Feel free to commit into this/doc branch as well if you want to fix some small detail or something.
@wozniakjan @zroubalik

wozniakjan commented on pkg/util/value_by_path.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+// GetValueByPath retrieves a value from a nested map using a dot-separated path
+func GetValueByPath(data map[string]interface{}, path string) (interface{}, error) {
```
suggestion: would you consider using `unstructured` helpers from kubernetes/apimachinery? There is a [`NestedFieldCopy`](https://github.com/kubernetes/apimachinery/blob/d794766488ac2892197a7cc8d0b4b46b0edbda80/pkg/apis/meta/v1/unstructured/helpers.go#L39-L45) function which is used for walking through yaml data, similar to the intent here. Given KEDA already depends on kubernetes/apimachinery, it might be good to reuse the exported code if it suits well.

wozniakjan replied:
thanks for checking, I agree with your findings

wozniakjan commented on pkg/scalers/metrics_api_scaler.go:
```diff
@@ -211,23 +246,126 @@ func parseMetricsAPIMetadata(config *scalersconfig.ScalerConfig) (*metricsAPISca
+		}
+		return r.AsApproximateFloat64(), nil
+	default:
+		return 0, fmt.Errorf("valueLocation must point to value of type number or a string representing a Quantity got: '%s'", v)
```
nit:
```suggestion
		return 0, fmt.Errorf(valueLocationWrongErrorMsg, v)
```

wozniakjan commented on pkg/scalers/metrics_api_scaler_test.go:
```diff
@@ -123,42 +123,34 @@ func TestMetricsAPIGetMetricSpecForScaling(t *testing.T) {
+		expectVal float64
+		expectErr bool
+	}{
+		{name: "integer", input: []byte(`{"components":[{"id": "82328e93e", "tasks": 32, "str": "64", "k":"1k","wrong":"NaN"}],"count":2.43}`), key: "count", format: JSONFormat, expectVal: 2.43},
```
nit: given the input repeats, can you please create a `var` or `const` for it?

```go
var rawJsonInput := []byte(`{"components":[{"id": "82328e93e", "tasks": 32, "str": "64", "k":"1k","wrong":"NaN"}],"count":2.43}`)
```

Friedrich42 replied:
done

wozniakjan commented on pkg/scalers/metrics_api_scaler_test.go:
```diff
@@ -123,42 +123,34 @@ func TestMetricsAPIGetMetricSpecForScaling(t *testing.T) {
+		{name: "{}.[].{}", input: []byte(`{"components":[{"id": "82328e93e", "tasks": 32, "str": "64", "k":"1k","wrong":"NaN"}],"count":2.43}`), key: "components.0.tasks", format: JSONFormat, expectVal: 32},
+		{name: "invalid data", input: []byte(`{"components":[{"id": "82328e93e", "tasks": 32, "str": "64", "k":"1k","wrong":"NaN"}],"count":2.43}`), key: "components.0.wrong", format: JSONFormat, expectErr: true},
+		{name: "integer", input: []byte(`{components: [{id: 82328e93e, tasks: 32, str: '64', k: 1k, wrong: NaN}], count: 2.43}`), key: "count", format: YAMLFormat, expectVal: 2.43},
+		{name: "string", input: []byte(`{components: [{id: 82328e93e, tasks: 32, str: '64', k: 1k, wrong: NaN}], count: 2.43}`), key: "components.0.str", format: YAMLFormat, expectVal: 64},
```
shouldn't the `input` here be in YAML?

Friedrich42 replied:
I've put it into a variable to avoid duplication

wozniakjan replied:
ah, I got so used to the non-compact YAML notation that I didn't realize the compact one can be this similar to JSON, thanks :)

wozniakjan commented on tests/scalers/metrics_api/metrics_api_test.go:
```diff
@@ -162,9 +162,10 @@ spec:
-      authMode: "basic"
-      method: "query"
+      valueLocation: value
+      format: json
```
does this mean users will need to default already existing metrics-api triggers with `format`? We should strive towards a code where the tests pass without this change imho.

Friedrich42 replied:
I've reverted that change

Friedrich42 commented:
@wozniakjan
Thank you for the review!

I've addressed most of the points.
Using k8s package is a great idea, but it may not be suitable here, because it doesn't support `.number` notation.

What do you think we should do about that?

wozniakjan commented:
> Using k8s package is a great idea, but it may not be suitable here, because it doesn't support .number notation.
> What do you think we should do about that?

I didn't realize the incompatibility, thanks for the thorough testing and you got me convinced to go with `GetValueByPath` :)

Now FOSSA tests appear to be failing, let me go through the repo to see if there is any info regarding these
```
Error:  An issue occurred
  *** Relevant Errors ***
      Error: The build failed. Check the FOSSA webapp for more details
      → src/App/Fossa/API/BuildWait.hs:119:41
```

Friedrich42 commented:
> > Using k8s package is a great idea, but it may not be suitable here, because it doesn't support .number notation.
> > What do you think we should do about that?
> 
> I didn't realize the incompatibility, thanks for the thorough testing and you got me convinced to go with `GetValueByPath` :)
> 
> Now FOSSA tests appear to be failing, let me go through the repo to see if there is any info regarding these
> 
> ```
> Error:  An issue occurred
>   *** Relevant Errors ***
>       Error: The build failed. Check the FOSSA webapp for more details
>       → src/App/Fossa/API/BuildWait.hs:119:41
> ```

I don't think it's actually failing, just need to restart it, because the itself report is clean

wozniakjan reviewed (approved):
lgtm
(not sure if approval from me is sufficient, maybe @kedacore/keda-contributors will need to approve as well)

JorTurFer commented:
/run-e2e metrics
